### ISSUE TITLE
Very slow initial data import

### ISSUE THREAD

**Issue by sirex** [atviriduomenys/spinta#473]
Currently `spinta push` command, is very slow, because it checks many things, to ensure data quality. But in some cases, for very large data imports, there should be a faster way, to import data, with relaxed checks.

One option is to use [PostgreSQL COPY](https://www.postgresql.org/docs/current/sql-copy.html) directly, without doing any checks, this should significantly increase speed at the cost of data quality checks. And this would only work on an empty dataset, assuming that all objects are new. This will also work only on a whole dataset, there would be no possibility to continue if something went wrong, on an error, all the data must be wiped out and imported again from scratch.

### Manual data import

Note!  This is a temporary solution, before automated solution is implemented.

For this to work, data should be provided as CSV files, each data table (model) must have two CSV files, one for the data and the other for changes.

Data table must have following columns:

- `_txn` - a UUIDv4 string, this is a transaction number and should be same UUID for all data, because all the data will be imported in one transaction.
- `_created` - should be datetime string when data were exported, for example `2023-06-01 08:52:41.775208`.
- `_updated` - should be empty string (NULL).
- `_id` - a UUIDv4 string, must be unique for each table and should be also present in keymap and push state tables.
- `_revision` - a UUIDv4 string, must be present in push state.
- other user defined data columns, that do not start with `_`.

Changelog table must have following columns:

- `_txn` - same as in data table.
- `_id` - monotonically increasing positive integer, currently limited up to 9,223,372,036,854,775,807.
- `_revision` - same as in data table.
- `_rid` - same as in `_id` in data table, an UUIDv4 unique identifier of data object.
- `datetime` - same as `_created` in data table.
- `action` - `insert` string
- `data` - JSON serialized data, excluding properties starting with `_`.

```sql
COPY "datasets/gov/lsd/cl/sektoriai/Sektorius" (
    _txn,
    _created,
    _updated,
    _id,
    _revision,
    kodas,
    pavadinimas_lt
)
FROM '/tmp/data.csv' DELIMITER ',' CSV HEADER;
```

For this to work, `/tmp/data.csv` file must be uploaded to the database server.

For this, we need a way, how to securely upload large CSV files and be able to import these files directly into database.

### Automation

In order to make this more convenient for end users, we can ask users to prepare just the JSON data file, which would be uploaded as file, and then imported using COPY command.

Before importing, JSON file will be transformed into two files, by adding all the _* reserved columns and generating changelog table content.

For this to work, we need to add an API endpoint, that would support uploading JSON files directly in a similar way as was described in #475.

First we need to create import job transaction by providing metadata about a file being uploaded:

Request:
```http
POST /datasets/gov/example/City/:import HTTP/1.1
Host: get.data.gov.lt
Content-Type: application/json

{
  "action": "create",
  "format": "jsonl",
}
```

- `action` - importavimo veiksmas
  - `create` - import on an empty table, returns error if table is not empty
  - `append` - import on existing table, appending new rows, only inserts are supported
  - `update` - import on existing table and update existing objects (this is reserved for future)https://github.com/atviriduomenys/prieziura/issues/118
- `format` - import file format, corresponds to an external backend of the same name

This creates a new transaction for data import and redirects to the import transaction endpoint:

Response:
```http
HTTP/1.1 308 Resume Incomplete
Location: /_txn/4d14d1d4-94b3-4f01-9634-e32e16769dae
```

Upload status could be checked using following API call:

Request:
```http
GET /_txn/4d14d1d4-94b3-4f01-9634-e32e16769dae HTTP/1.1
```

Response:
```http
HTTP/1.1 200 OK

{
  "txn": "4d14d1d4-94b3-4f01-9634-e32e16769dae",
  "path": "datasets/gov/example/Model",
  "created": "2024-08-27 11:17:12.192208",
  "status": "uploading",
  "format": "jsonl",
  "chunks": 0,
}
```

- `txn` - import transaction id
- `path` - comes from `POST` path and specified location, where data should be uploaded, client must have write access to namespace or model specified in path, if path is a namespace, then uploaded data might contain data for multiple models, in this case `_type` must be provided with each data object
- `model` - model name into which data should be uploaded, `ns` or `model` must be specified, but not both
- `created` - when transaction was created
- `status` - import status, possible options
  - `error` - There was an error, provided in `errors` list.
  - `uploading` - Incomplete upload, if chunked upload is used.
  - `queued` - Uploaded file is waiting in queue for processing.
  - `checking` - Uploaded file is being checked and transformed into files ready to be imported.
  - `importing` - Uploaded file is being imported into database.
  - `complete` - Import is complete, transaction is closed.
- `format` - import file format, corresponds to an external backend of the same name
- `chunks` - number of uploaded chunks.

After transaction is created, we can start to upload data, optionally using `Content-Range`:

Request:
```http
PATCH /_txn/4d14d1d4-94b3-4f01-9634-e32e16769dae HTTP/1.1
Host: get.data.gov.lt
Content-Type: application/jsonl
Content-Range: lines 0-1000/*
Content-Length: 1000

{"column1": "value1", "column2": "value2"}
...
```

Response:
```http
HTTP/1.1 206 Partial Content
Location: /_txn/4d14d1d4-94b3-4f01-9634-e32e16769dae

{
  "txn": "4d14d1d4-94b3-4f01-9634-e32e16769dae",
  "created": "2024-08-27 11:17:12.192208",
  "status": "uploading",
  "chunks": 1,
}
```

When last bytes are uploaded, we return  `200 OK` instead of `206 Partial Content` and import status is changed to `queued`.

Request:
```http
PATCH /_txn/4d14d1d4-94b3-4f01-9634-e32e16769dae HTTP/1.1
Host: get.data.gov.lt
Content-Type: application/jsonl
Content-Length: 0
```

Response:
```http
HTTP/1.1 200 OK
Location: /_txn/4d14d1d4-94b3-4f01-9634-e32e16769dae

{
  "txn": "4d14d1d4-94b3-4f01-9634-e32e16769dae",
  "created": "2024-08-27 11:17:12.192208",
  "status": "queued"
  "chunks": 123,
}
```

### Implementation details

Under the hood, for each import job a directory should be created under configured upload root (`direct_upload_dir` config option):

```python
txn[:2]/txn[2:]/
```

This directory should contain following files:

- `data/{uuid}.jsonl.gz` - original user uploaded data files, each chunk stored as a separate gzipped file (use [gzip](https://docs.python.org/3/library/gzip.html))
- `status.json` - all the information about transaction, basically this file is returned on `get /_txn/{id}` request

There should be a background service, which would do processing of uploaded files.

Each uploaded chunk file, should be transformed into files, that can be imported directly into PostgreSQL. One of the options is to convert JSON Lines chunk files into two CSV files:

- `import/{dataset}/{model}/data.csv`
- `import/{dataset}/{model}/changes.csv`

Currently we could start with simple check, like CSV header if it matches manifest file, but in future, would be nice, to do full check using Dask, but this should be a separate task.

### Required scopes

Since we are introduction a new `:import` action, same authorization scope rules applies.

Client can have a global import scope for example:

    spinta_import

Or a namespace, dataset or model specific scope, for example:

    spinta_datasets_gov_ivpk_adp_import
    spinta_datasets_gov_ivpk_adp_catalog_import
    spinta datasets_gov_ivpk_adp_catalog_model_import

See: https://spinta.readthedocs.io/en/latest/manual/access.html#scopes

### `spinta importer`

`spinta importer` is a CLI command, that runs importer service, this service is responsible for monitoring import tasks.

Importer should monitor all import jobs in `direct_upload_dir` directory, with status `queued`. When a `queued` job is found, then importer does following things:

1. Changes status to `checking` and runs some checks.
2. Changes status to `importing` and starts data import process into data backend.
3. Changes status to `complete`, once import is finished.

Importer should remember job statuses and should skip completed jobs.

Importer should be run in a single process, concurrent importers should not be allowed.  For this probably we need to add a lock file in `direct_upload_dir`, and prevent concurrent importers to be run. Multiple importers could run at the same time if different namespaces are specified (`--ns {namesace}` argument).

CLI user interface:

`spinta importer` - prints job statuses, excluding completed jobs.

`spinta importer --import` - run through all jobs and do import on all queued jobs if any, exit when done.

`spinta importer --import txn` - import a specific job, if this job is in queue.

`spinta importer --daemon` - run an infinite loop and run through all jobs, to check if there are any changes in statuses, waiting for `queued` status.

Other importer command line arguments:

`spinta -o direct_upload_dir=` - use standard Spinta configuration options handling.

`-i, --import` do a single import round
`-i, --import {txn}` import a single job (status must be `queued`)
`-d, --daemon` run infinite loop and do continuous imports
`-n, --ns, --namespace {namspace}` run imports only on a specific namespace, this way we can have multiple importers running on separate namespaces 

### Deleting completed jobs

All completed jobs should be deleted after `direct_upload_retention` seconds of time.  By default files should be deleted after 1 week.

### Import logs

All job status changes should be logged into `{direct_upload_dir}/log.json` file, where single line should  look like this (example bellow is split into multiple lines for readability):

```json
{
  "time": "2024-07-31T07:37:03.552142+00:00",
  "txn": "4d14d1d4-94b3-4f01-9634-e32e16769dae",
  "status": "uploading",
  "chunk": "9051c4bc-0121-4759-9e61-e4bbc4fc67cf"
}
```

### `spinta push` integration

This should be done as a separate task, to integrate direct upload functionality with `spinta push` command.

If push target is empty, then direct file upload should be preferred, using 100M ranges (not sure about 100M, best performing and accepted by the HTTP servers value should be used).

`push` command should show progress bar, when uploading, and then wait for other transaction statuses, until data import is fully complete.

If data on target already exists, then current push mechanism should be used.

On initial direct file upload, on completion, keymap and push state should be synced with target. Probably we don't need to sync, just calculate correct values and initialized keymap ant push state tables, with these values.

### Related

- https://github.com/atviriduomenys/spinta/issues/475
- https://github.com/atviriduomenys/spinta/issues/572
- https://github.com/atviriduomenys/spinta/issues/960
- https://github.com/atviriduomenys/prieziura/issues/118

**Comment by sirex** [comment]
Currently data import is very slow, because we do data validation, integrity checks and write data using separate query for each object, for very large initial imports, this could take days.

This issue is mainly related with:

- initial data imports
- data migrations, when new columns are added or whole columns have to be updated

For such cases, we need a separate action:

    /ns/Model/:import

Which would accept JSON data as a file, and would be able to upload this file and import it as a whole in a fastest possible way.

**Comment by JustinasKen** [comment]
There are some questions I have for implementation:

1. What is the actual endpoint for this kind of data import. In the comment above you mentioned `/:import`, but in previous examples, they were simple `POST`/`PATCH` requests.
2. You have mentioned, that there should be a separate background process that would validate uploaded files, can you go into more details, on how exactly that should work.
3. I assume you will going to need specific permissions in order to upload these files, if so, what are they going to be called ?
4. What will happen to uploaded files over longer period of time (each transaction will create copies of files) ?
5. I am confused about, how will `changelog` and `data` will be stored in the same `CSV` file, since `changelog` can store multiple entries about a single row. Why not upload 2 separate files ?
6. I assume that `spinta push` integration is not currently needed and also, how would we know which file to use, since they are all stored under transactions ?
7. In previous comment you have mentioned that it should accept `JSON` data as a file, but aren't data supposed to be stored in `CSV` file ?
8. You have mentioned that it should also support compressed `CSV`, can you mention what format we should accept ?

**Comment by sirex** [comment]
> 1. What is the actual endpoint for this kind of data import. In the comment above you mentioned /:import, but in previous examples, they were simple POST/PATCH requests.

Updated description, `/:import` should be used, to distinguish import from other normal data operations.

> 2. You have mentioned, that there should be a separate background process that would validate uploaded files, can you go into more details, on how exactly that should work.

Added separate section in the description for `spinta importer`.

> 3. I assume you will going to need specific permissions in order to upload these files, if so, what are they going to be called ?

`spinta_{namespace}_import`

Also added a separate section called "Rquired scopes".

> 4. What will happen to uploaded files over longer period of time (each transaction will create copies of files) ?

Each transaction accepts new uploaded files. Once files are imported and transaction is completed, transaction directory with all files in it should be deleted. All status changes, including transaction file deletion should be added to a log.

> 5. I am confused about, how will changelog and data will be stored in the same CSV file, since changelog can store multiple entries about a single row. Why not upload 2 separate files ?

In this issue description we provide to options:

1. **Manual data import** - in this case, client must provide two CSV files, one for data, other for changes, basically everything PostgreSQL expects. This is a temporary solution, until there is no automation.
2. **Automation** - in this cases, use provides only one data CSV file, without `_id`, `_revision` and other metadata.

Only **Automation** parts requires changes to Spinta code. In automated data import via API, client provides only data file, and Spinta does following:

1. Simply uploads whatever data is passed through API and stores it to a file.
2. Separate importer process will process uploaded files, by creating separate data file and changelog, adding metadata fields liek `_id` and `_revision` if not provided and does everything else, what is needed in order to import data to a backend.

> 6. I assume that spinta push integration is not currently needed and also, how would we know which file to use, since they are all stored under transactions ?

`spinta push` integration should be a separate task. I suggest, to create separate task, for all work needed and leave this as an epic. Because it might require multiple separate tasks..

I did not understood last part of the question. Could you clarify it?

> 7. In previous comment you have mentioned that it should accept JSON data as a file, but aren't data supposed to be stored in CSV file ?

For **Manual data import** user provides CSV file, which matches table structure in PostgreSQL database identically.

For **Automated import** user provides JSON file, using same structure as using normal API. The `/:import` action accepts JSON file as with normal API, but instead of importing this file   on request, postpones importing and file processing after file is uploaded.

> 8. You have mentioned that it should also support compressed CSV, can you mention what formats we should accept ?

I think we should not expose internal specific backend structure to the client, and instead of PostgreSQL compatible CSV file, a JSON Lines file should be uploaded. I fixed that in the description.

Also, made other additions in the task description, like saving separate chunks, in separate files for better reliability and to avoid issues with concurrent writes.

Also I have a feeling, that we might use Kafka instead of inventing a custom protocol.

**Comment by Grigorjeva11** [comment]
Reikia komentaro, kaip turės veikti ir kokie niuansai bus po architekturos pakeitimo @LaurynasVenck @sirex

**Comment by JuliusLADP** [comment]
Sugeneruoti du failus (pradinį ir pakeitimų) lokaliai išsaugant failus, tokius Ckurie yra tinkami sukelti į spintą

**Comment by sirex** [comment]
Consider using columnar DuckDB backend for fast import/export operations on analytical data.

- https://github.com/atviriduomenys/spinta/issues/572

DuckDB supports SQL, so it can be used with existing sql backend.

**Comment by Grigorjeva11** [comment]
Laukiam sprendimo dėl architekturos

**Comment by SauliusPetraitis** [comment]
Laukia architektūrinių sprendimų + pasižiūrėti reikia jei pagerės, tai kiek dėl DB cluster.
